core/src/main/scala/com/microsoft/azure/synapse/ml/causal/LinearDMLEstimator.scala

```diff
@@ -0,0 +1,279 @@
+            Please select a treatment model that supports sample weight.""".stripMargin)
+        }
+        getOutcomeModel match {
+          case w: HasWeightCol => w.set(w.weightCol, getWeightCol)
+          case _ => throw new Exception("""The selected outcome model does not support sample weight,
+            but the weightCol parameter was set for the LinearDMLEstimator.
+            Please select a outcome model that supports sample weight.""".stripMargin)
+        }
+      }
+
+      // sampling with replacement to redraw data and get TE value
+      // Run it for multiple times in parallel, get a number of TE values,
+      // Use average as Ate value, and 2.5% low end, 97.5% high end as Ci value
+      // Create execution context based on $(parallelism)
+      log.info(s"Parallelism: $getParallelism")
+      val executionContext = getExecutionContextProxy
+
+      val ateFutures =(1 to getMaxIter).toArray.map { index =>
+        Future[Option[Double]] {
+          log.info(s"Executing ATE calculation on iteration: $index")
+          // If the algorithm runs over 1 iteration, do not bootstrap from dataset, otherwise, draw sample with replacement
+          val redrewDF =  if (getMaxIter == 1) dataset else dataset.sample(withReplacement = true, fraction = 1)
+          val ate: Option[Double] =
+            try {
+              val totalTime = new StopWatch
+              val oneAte = totalTime.measure {
+                trainInternal(redrewDF)
+              }
+              log.info(s"Completed ATE calculation on iteration $index and got ATE value: $oneAte, time elapsed: ${totalTime.elapsed() / 60000000000.0} minutes")
+              Some(oneAte)
+            } catch {
+              case ex: Throwable =>
+                log.warn(s"ATE calculation got exception on iteration $index with the redrew sample data. Exception details: $ex")
+                None
+            }
+          ate
+        }(executionContext)
+      }
+
+      val ates = awaitFutures(ateFutures).flatten
+      val dmlModel = this.copyValues(new LinearDMLModel(uid)).setAtes(ates.toArray)
+      dmlModel
+    })
+  }
+
+  private def trainInternal(dataset: Dataset[_]): Double = {
+
+    def getModel(model: Estimator[_ <: Model[_]], labelColName: String, excludedFeatures: Array[String]) = {
+      model match {
+        case classifier: ProbabilisticClassifier[_, _, _] => (
+          new TrainClassifier()
+            .setModel(model)
+            .setLabelCol(labelColName)
+            .setExcludedFeatures(excludedFeatures),
+          classifier.getProbabilityCol,
+          Seq(classifier.getPredictionCol, classifier.getProbabilityCol, classifier.getRawPredictionCol)
+        )
+        case regressor: Regressor[_, _, _] => (
+          new TrainRegressor()
+            .setModel(model)
+            .setLabelCol(labelColName)
+            .setExcludedFeatures(excludedFeatures),
+          regressor.getPredictionCol,
+          Seq(regressor.getPredictionCol)
```

Comment:
this code looks like it returnns Seq to get around not returning some things in the second branch. I would make this function generic to any SparkML model, and just return the moidel. Then later when you need to determine if you have a prob and a raw prediction column later in your code, pattern match on the model and use the getters.

Reply:
DML estimator's treatment/outcome model only works with "Regressor" and "ProbabilisticClassifier", this is an internal function for TrainInternal method, and it contains some domain knowledge, we use classifier's probability and regression's prediction column to calculate residual.

Reply:
> this code looks like it returnns Seq to get around not returning some things in the second branch. I would make this function generic to any SparkML model, and just return the moidel. Then later when you need to determine if you have a prob and a raw prediction column later in your code, pattern match on the model and use the getters.

Good point. I'll discuss it with Dylan.

Reply:
Also @memoryz, i now see these two models are just wrappers around the actual models so the need to make this more generic from a model standpoint inst as important. The refactor and pattern match though might still be nice to do